packages/zoe/src/contracts/priceAggregator.js

```diff
@@ -0,0 +1,334 @@
+import { makeNotifierKit } from '@agoric/notifier';
+import makeStore from '@agoric/store';
+import { assert, details } from '@agoric/assert';
+import {
+  calculateMedian,
+  natSafeMath,
+  makeOnewayPriceAuthorityKit,
+} from '../contractSupport';
+
+import '../../exported';
+import './exported';
+
+const { add, multiply, floorDivide, ceilDivide, isGTE } = natSafeMath;
+
+/**
+ * This contract aggregates price values from a set of oracles and provides a
+ * PriceAuthority for their median.
+ *
+ * @type {ContractStartFn}
+ */
+const start = async zcf => {
+  const {
+    timer: rawTimer,
+    POLL_INTERVAL,
+    maths: { In: mathIn, Out: mathOut },
+    baseAmountIn = mathIn.make(1),
+  } = zcf.getTerms();
+
+  const baseValueIn = mathIn.getValue(baseAmountIn);
+
+  /** @type {TimerService} */
+  const timer = rawTimer;
+
+  /** @type {IssuerRecord & { mint: ERef<Mint> }} */
+  let quoteKit;
+
+  /** @type {PriceAuthority} */
+  let priceAuthority;
+
+  /** @type {PriceAuthorityAdmin} */
+  let priceAuthorityAdmin;
+
+  /** @type {number} */
+  let publishedBaseValueOut;
+
+  /**
+   *
+   * @param {PriceQuoteValue} quote
+   */
+  const authenticateQuote = async quote => {
+    const quoteAmount = quoteKit.amountMath.make(harden(quote));
+    const quotePayment = await E(quoteKit.mint).mintPayment(quoteAmount);
+    return harden({ quoteAmount, quotePayment });
+  };
+
+  const { notifier, updater } = makeNotifierKit();
+  const zoe = zcf.getZoeService();
+
+  /**
+   * @typedef {Object} OracleRecord
+   * @property {(timestamp: Timestamp) => Promise<void>=} querier
+   * @property {number} lastSample
+   */
+
+  /** @type {Set<OracleRecord>} */
+  const oracleRecords = new Set();
+
+  /** @type {Store<Instance, Set<OracleRecord>>} */
+  const instanceToRecords = makeStore('oracleInstance');
+
+  let publishedTimestamp = await E(timer).getCurrentTimestamp();
+
+  // Wake every POLL_INTERVAL and run the queriers.
+  const repeaterP = E(timer).createRepeater(0, POLL_INTERVAL);
+  /** @type {TimerWaker} */
+  const waker = {
+    async wake(timestamp) {
+      // Run all the queriers.
+      const querierPs = [];
+      oracleRecords.forEach(({ querier }) => {
+        if (querier) {
+          querierPs.push(querier(timestamp));
+        }
+      });
+      await Promise.all(querierPs);
+    },
+  };
+  E(repeaterP).schedule(waker);
+
+  /**
+   * @param {Object} param0
+   * @param {number} [param0.overrideBaseValueOut]
+   * @param {Timestamp} [param0.timestamp]
+   */
+  const makeCreateQuote = ({ overrideBaseValueOut, timestamp } = {}) =>
+    /**
+     * @param {PriceQuery} priceQuery
+     * @returns {ERef<PriceQuote>=}
+     */
+    function createQuote(priceQuery) {
+      // Sniff the current baseValueOut.
+      const baseValueOut =
+        overrideBaseValueOut === undefined
+          ? publishedBaseValueOut
+          : overrideBaseValueOut;
+      if (baseValueOut === undefined) {
+        return undefined;
```

Comment:
I don't get what baseValueOut and baseValueIn are being used for. What's their role here?

Reply:
`baseValueIn` divided by `baseValueOut` is the current aggregated price, used as the scaling factor for calculating prices of other values of `valueIn` or `valueOut`.

Reply:
The aggregated oracles need to provide quote data of how much `brandOut` you would get for `baseAmountIn`.

Reply:
I renamed `baseValueOut` to `valueOutForBaseValueIn`?